node/src/protocol/presignature.rs

```diff
@@ -0,0 +1,228 @@
+
+/// Abstracts how triples are generated by providing a way to request a new triple that will be
+/// complete some time in the future and a way to take an already generated triple.
+pub struct PresignatureManager {
+    /// Completed unspent presignatures.
+    presignatures: HashMap<PresignatureId, Presignature>,
+    /// Ongoing triple generation protocols.
+    generators: HashMap<PresignatureId, PresignatureGenerator>,
+
+    participants: Vec<Participant>,
+    me: Participant,
+    threshold: usize,
+    epoch: u64,
+}
+
+impl PresignatureManager {
+    pub fn new(
+        participants: Vec<Participant>,
+        me: Participant,
+        threshold: usize,
+        epoch: u64,
+    ) -> Self {
+        Self {
+            presignatures: HashMap::new(),
+            generators: HashMap::new(),
+            participants,
+            me,
+            threshold,
+            epoch,
+        }
+    }
+
+    /// Returns the number of unspent presignatures available in the manager.
+    pub fn len(&self) -> usize {
+        self.presignatures.len()
+    }
+
+    /// Returns the number of unspent presignatures we will have in the manager once
+    /// all ongoing generation protocols complete.
+    pub fn potential_len(&self) -> usize {
+        self.presignatures.len() + self.generators.len()
+    }
+
+    /// Starts a new presignature generation protocol.
+    pub fn generate(
+        &mut self,
+        triple0: Triple,
+        triple1: Triple,
+        public_key: &PublicKey,
+        private_share: &PrivateKeyShare,
+    ) {
+        let id = rand::random();
+        tracing::info!(id, "starting protocol to generate a new presignature");
+        let protocol = Box::new(
+            cait_sith::presign(
+                &self.participants,
+                self.me,
+                PresignArguments {
+                    triple0: (triple0.share, triple0.public),
+                    triple1: (triple1.share, triple1.public),
+                    keygen_out: KeygenOutput {
+                        private_share: *private_share,
+                        public_key: *public_key,
+                    },
+                    threshold: self.threshold,
+                },
+            )
+            .unwrap(),
+        );
+        self.generators.insert(
+            id,
+            PresignatureGenerator {
+                protocol,
+                triple0: triple0.id,
+                triple1: triple1.id,
+            },
+        );
+    }
+
+    /// Ensures that the presignature with the given id is either:
+    /// 1) Already generated in which case returns `None`, or
+    /// 2) Is currently being generated by `protocol` in which case returns `Some(protocol)`, or
+    /// 3) Has never been seen by the manager in which case start a new protocol and returns `Some(protocol)`, or
+    /// 4) Depends on triples (`triple0`/`triple1`) that are unknown to the node
+    // TODO: What if the presignature completed generation and is already spent?
```

Comment:
Also, can it happen that presignature generation is completed but it was not spent?

Reply:
Yes, that's the first case: `1) Already generated in which case returns None`